### PR TITLE
Fix blinking polyline

### LINKED ISSUE #6741: Polyline artifacts depending on camera position
In master I'm seeing new artifacts when rendering polylines.  Possibly this is related to the polylines on terrain addition?

[Sandcastle](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=ZVFdT8IwFP0rzV42EuwYxI/gJCIagzHxYcQH05fSXaSha0k/IMP4371jiKhNH3o/zuk59264JRsJW7DkhmjYkgk4GSr6us8lsdiHE6M9lxps3CUfTBM8bmmCKsdaVtwDGZIFVw6Y/uxcM810S0lL7nlhgkUWyssyOZBPdpW6P5aoMrxMYkrT4x2v1y4teLVW0PSlTjZPKhAXdzrUL0Eni6CFl0YnP5909uJafQcF3nKxgvJBe+lrtPjTTKHJSVT2Dv6unqKEAq0oJT2kj2Cghixu3ZzwCWXEiopgLcJnsgLkPLh6CkpyjXqBLqypps5cXfSyJO73sv5Zb3CWnc+y3rB/ORz035D4oPM4MqajbpQ7XysYtaVbdG2sJ8GqBIfiAYeA7C6dB/TkqXCugebpNygv5YbI8oZFf9bGIiIUdw4ri6BUIXfAolGeYv8vWLMJqd9fNmAVr5uWZTZ6bpOU0jzF8D/KG6Pm3J4wMv0F)

note: sandcastle is the hosted version which does not have the problem.  copy/paste the code locally

capture

### PR DESCRIPTION
**Pull request by bagnell**
Fixes #6741.

Fixes this https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issuecomment-404742328
but not https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issuecomment-406717425
or https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issuecomment-408198164
which are separate issues.

This is probably a workaround for an issue where a segment of the polyline intersects the near plane and log depth is enabled.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @bagnell!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/fix-blinking-polyline/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by hpinkos** [comment]
I can confirm this fixes the problem.  Will subdividing the line at a higher frequency have an impact on performance?  I'm not sure it's worth putting in this workaround because users can already workaround the issue by setting `scene.logarithmicDepthBuffer: false`

**Comment by ggetz** [comment]
@bagnell @hpinkos So whats the plan here? If we have a fix, it should be merged today.

**Comment by pjcozzi** [comment]
> Fixes this #6741 (comment)
but not #6741 (comment)
or #6741 (comment)
which are separate issues.

@bagnell please submit separate issues for those.

File: Source/Core/PolylinePipeline.js [hunk 1]
```diff
@@ -309,7 +309,79 @@ define([
...
+        }
+        return newPositions;
+    };
+
+    var distanceScratch = new Cartesian3();
+    function getPointAtDistance(p0, p1, distance, length) {
+        Cartesian3.subtract(p1, p0, distanceScratch);
+        Cartesian3.multiplyByScalar(distanceScratch, distance / length, distanceScratch);
+        Cartesian3.add(p0, distanceScratch, distanceScratch);
+        return [distanceScratch.x, distanceScratch.y, distanceScratch.z];
```
Comment (pjcozzi) [inline Source/Core/PolylinePipeline.js]: Can this function be rewritten to take a scratch `Cartesian3` and return it?

File: Source/Core/PolylineGeometry.js [hunk 2]
```diff
@@ -361,6 +362,12 @@ define([
                 ellipsoid: ellipsoid,
                 height: heights
             });
+        } else {
+            // This subdivides a line at a much greater granularity. This may be a workaround for an actual bug
+            // where a polyline blinks in and out during camera movement when log depth is enabled and a segment of
+            // the polyline intersects the near plane.
+            minDistance = CesiumMath.chordLength(granularity * 10.0, ellipsoid.maximumRadius);
```
Comment (pjcozzi) [inline Source/Core/PolylineGeometry.js]: I don't have 100% context here, but shouldn't this ignore `granularity` and just pick a min distance to subdivide, which could be as simple as `ellipsoid.maximumRadius`?  Previous behavior was that `granularity` was ignored unless the polyline follows the surface - why would that change now - when would a user want to explicitly provide `granularity` for non-surface lines - it doesn't even make sense as a measurement for arbitrary lines, right?

**Comment by pjcozzi** [comment]
Update CHANGES.md.

**Comment by hpinkos** [comment]
@bagnell are we trying to get this in for the release?

**Comment by bagnell** [comment]
This is updated and ready for another review.

**Comment by lilleyse** [comment]
I can confirm this fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issuecomment-404742328.

Is this intended to fix https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issue-336373649? I still see the polyline blink out occasionally.

@bagnell who is the best person to do a follow up review?

**Comment by hpinkos** [comment]
@bagnell merge in master

**Comment by bagnell** [comment]
> Is this intended to fix [#6741 (comment)](https://github.com/AnalyticalGraphicsInc/cesium/issues/6741#issue-336373649)? I still see the polyline blink out occasionally.

I can't reproduce that issue in this branch or in master.

**Comment by bagnell** [comment]
> who is the best person to do a follow up review?

Either, you or @hpinkos.

**Comment by hpinkos** [comment]
yeah, this change looks fine to me.  My only concern was it's impact on performance

**Comment by hpinkos** [comment]
Since this is a change to `PolylineGeometry` and not `Polyline` I'm sure it's fine though

**Comment by lilleyse** [comment]
> Is this intended to fix #6741 (comment)? I still see the polyline blink out occasionally.
> > I can't reproduce that issue in this branch or in master.

I can reproduce it on one machine but not another. Maybe it's a driver difference since the flickery one is on AMD and the other is Nvidia. However jumping back to a random older version like 1.34 I don't see the problem on either.

**Comment by bagnell** [comment]
@lilleyse what do you want to do here? I can't reproduce the issue and I don't have an AMD card.

**Comment by lilleyse** [comment]
Maybe merge this but keep https://github.com/AnalyticalGraphicsInc/cesium/issues/6741 open and remove the `priority next release` label.

**Comment by lilleyse** [comment]
Or open a new issue because that one is really crowded. Anyways I'll merge once CI passes.

**Comment by mramato** [comment]
@lilleyse is there a reason this was never merged?  It's marked next release.

**Pull request merged**